rust_src/src/process.rs

```diff
@@ -263,6 +268,32 @@ pub fn process_status(process: LispObject) -> LispObject {
     status
 }
 
+/// Return the connection type of PROCESS.
+/// The value is either the symbol `real', `network', `serial', or `pipe'.
+/// PROCESS may be a process, a buffer, the name of a process or buffer, or
+/// nil, indicating the current buffer's process.
+#[lisp_fn]
+pub fn process_type(process: LispObject) -> LispObject {
+    let mut p;
+
+    if process.is_string() {
+        p = get_process(process);
+        if p.is_nil() {
+            p = get_buffer_process(process);
+        }
+    } else {
+        p = unsafe { cget_process(process.to_raw()) }
+    };
+
+    if p.is_nil() {
+        return p;
+    }
+
+    let p_ref = p.as_process_or_error();
+
+    p_ref.process_type()
+}
```

Comment:
```
let process = if process.is_string() {
    let p = get_process(process);
    if p.is_nil() {
        get_buffer_process(process)
    } else {
        unsafe { cget_process(process) }  // no more to_raw
   }
};

if process.is_nil() {
    process
} else {
    let p_ref = process.as_process_or_error();
    p_ref.process_type()
}
```

Reply:
Overwriting the function argument with the transformed version of it is also a common idiom.

Reply:
The original code uses a single function called `get_process`. Why not port it using the code above as a beginning? I see there is a `get_process` in this Rust file already. You can rename it to `get_process_lisp` and set the name to be `get_process` in the lisp_fn macro so it still works in C.

Reply:
Great :) I thought the `mut` was weird, just distracted myself a bit with the nested conditionals.